### ISSUE TITLE
Error in config mapping and environment variables

### ISSUE THREAD

**Issue by mcanzerini** [quarkusio/quarkus#30657]
### Describe the bug

I defined a Map<String, String> field in a ConfigMapping interface like: 
```
@ConfigMapping(prefix="parameters")
public interface ConfigExample {

    ParametersExample parametersExample();

    interface ParametersExample {
        Map<String, String> mapExample();
    }

}
```

When I run my quarkus app in a docker context like: 
`docker run -ti -e parameters.parameters-example.map-example.TEST=env1 reproducer-image`

Quarkus does not start at all saying me: 

```
ERROR: Failed to start application (with profile [oc])
java.util.NoSuchElementException: SRCFG00014: The config property parameters.parameters-example.map-example.test is required but it could not be found in any config source
        at io.smallrye.config.SmallRyeConfig.convertValue(SmallRyeConfig.java:298)
        at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:242)
        at io.smallrye.config.ConfigMappingProvider.lambda$processLazyMapValue$0(ConfigMappingProvider.java:507)
        at io.smallrye.config.ConfigMappingProvider.mapConfiguration(ConfigMappingProvider.java:977)
        at io.smallrye.config.ConfigMappingProvider.lambda$mapConfiguration$3(ConfigMappingProvider.java:941)
        at io.smallrye.config.SecretKeys.lambda$doUnlocked$0(SecretKeys.java:20)
        at io.smallrye.config.SecretKeys.doUnlocked(SecretKeys.java:29)
        at io.smallrye.config.SecretKeys.doUnlocked(SecretKeys.java:19)
        at io.smallrye.config.ConfigMappingProvider.mapConfiguration(ConfigMappingProvider.java:941)
        at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:533)
        at io.quarkus.runtime.generated.Config.readConfig(Unknown Source)
        at io.quarkus.deployment.steps.RuntimeConfigSetup.deploy(Unknown Source)
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:108)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
        at org.acme.Main.main(Main.java:14)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
```
**This error message is really weird.**

If I run `docker run -ti -e parameters.parameters-example.map-example.test=env1 reproducer-image`

My app does start but obviously the Map is not populated with the key 'TEST'. 

**So the only workaround I found is to run:**
`docker run -ti -e parameters.parameters-example.map-example.test=env1 -e parameters.parameters-example.map-example.TEST=env1 reproducer-image`

I have tried to convert the property key in ABC_DEF format like the doc says:

If I run `docker run -ti -e parameters_parameters_example_map_example_TEST=env1 reproducer-image`
the application starts but the property env1 is not in the Map.

**Last but not least:** If I predefine this variable `parameters.parameters-example.map-example.PREDEFINED=value1` in my `application.properties` and I run `docker run -ti -e parameters_parameters_example_map_example_PREDEFINED=env1 reproducer-image` this works and the variables has been overridden with env1. 

I am really confused about that, especially with the first error message expecting the key with lower case.

Is this a bug or did I misunderstand something ?

### Expected behavior

The app starts properly and the value passed by the environment variable populates the Map

### Actual behavior

Depending on case, the app does not start or the app starts without populating the Map

### Quarkus version or git rev

2.16.0

### Additional information

If needed, I can provide you a small reproducer

**Comment by geoand** [comment]
@radcortez perhaps some of your latest fixes have already taken care of this?

**Comment by radcortez** [comment]
> **Last but not least:** If I predefine this variable `parameters.parameters-example.map-example.PREDEFINED=value1` in my `application.properties` and I run `docker run -ti -e parameters_parameters_example_map_example_PREDEFINED=env1 reproducer-image` this works and the variables has been overridden with env1.

This is how you should do it and it is our recommendation in the docs:
https://quarkus.io/version/main/guides/config-reference#environment-variables

Environment variables go through a conversion process to a dotted format, because most environment only support [A-Za-z] and `_`. The problem here is that your property name is confusing because it mixes environment variables, dotted format and uppercase, so the additional property definition helps the system to disambiguate.

**Comment by radcortez** [comment]
Let me think if I can find a way to do it. It may be tricky to detect such cases.

**Comment by armudgal** [comment]
@radcortez 
Were you able to think of a way that this can be better tackled?

**Comment by radcortez** [comment]
I've changed how env vars are handled, which should help, but the update will only be available in Quarkus 3.5.x.
